### PR TITLE
add subcommands enable and install addons for application manager

### PR DESCRIPTION
**Pull request by mikeshng**
- add new subcommand to install built-in addons. 
- add new subcommand to enable addons. 
- implement application-manager addon for install and enable.

usages:
```
clusteradm install addons --names application-manager # Deploys the built-in add-on application manager operators to the hub cluster
```
```
clusteradm enable addons --names application-manager --clusters cluster1,cluster2 # Deploys the ManagedClusterAddOn CRs to the given managed clusters
```

Signed-off-by: Mike Ng <ming@redhat.com>

### CODE REVIEW

**Comment by mikeshng** [comment]
/hold

Work in progress.

File: pkg/cmd/enable/addons/exec.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+//ClusterName: The cluster name used in the template
+type ClusterName struct {
+	ClusterName string
+}
+
+func (o *Options) complete(cmd *cobra.Command, args []string) (err error) {
+	klog.V(1).InfoS("addon options:", "dry-run", o.ClusteradmFlags.DryRun, "names", o.names, "clusters", o.clusters, "output-file", o.outputFile)
+
+	if err := o.validate(); err != nil {
```
Comment (itdove) [inline pkg/cmd/enable/addons/exec.go:27]: The validate is automatically called by the cmd definition. I think the lines below the validate should go in the run

File: pkg/cmd/enable/addons/exec.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	"k8s.io/klog/v2"
+
+	clusterclientset "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	"open-cluster-management.io/clusteradm/pkg/cmd/enable/addons/scenario"
+	"open-cluster-management.io/clusteradm/pkg/helpers"
+	"open-cluster-management.io/clusteradm/pkg/helpers/apply"
+)
+
+//ClusterName: The cluster name used in the template
+type ClusterName struct {
```
[the remark this answers is not in the document]
Comment (itdove) [reply pkg/cmd/enable/addons/exec.go:20]: but this is ok

[the remark this answers is not in the document]
Comment (mikeshng) [reply pkg/cmd/enable/addons/exec.go:20]: As discussed, there are multiple cluster names and I want to specifically render the cluster name I want inside that loop.

File: pkg/cmd/enable/addons/exec.go (hunk 1 above)
Comment (mikeshng) [reply pkg/cmd/enable/addons/exec.go]: done as suggested

File: pkg/cmd/enable/addons/exec.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	kubeClient, apiExtensionsClient, dynamicClient, err := helpers.GetClients(o.ClusteradmFlags.KubectlFactory)
+	if err != nil {
+		return err
+	}
+
+	applierBuilder := &apply.ApplierBuilder{}
+	applier := applierBuilder.WithClient(kubeClient, apiExtensionsClient, dynamicClient).Build()
+
+	for _, addon := range o.values.addons {
+		if addon == "application-manager" {
```
Comment (qiujian16) [inline pkg/cmd/enable/addons/exec.go:104]: I think  we may not install hub  controller here, or we need  to explicitly call out somewhere (maybe in the flags or help message of cmd?) that application-manager is a built-in addon.

Also I would think we need to split "enable addon on managed cluster" and "install addon" to two commands

File: pkg/cmd/enable/addons/exec.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	if o.clusters == "" {
+		return fmt.Errorf("clusters is misisng")
+	}
+
+	return nil
+}
+
+func (o *Options) run() error {
```
Comment (qiujian16) [inline pkg/cmd/enable/addons/exec.go:49]: should we have some integration test around this?

File: pkg/cmd/enable/addons/exec.go (hunk 3 above)
Comment (mikeshng) [reply pkg/cmd/enable/addons/exec.go:104]: done as recommended.

File: pkg/cmd/enable/addons/exec.go (hunk 4 above)
Comment (mikeshng) [reply pkg/cmd/enable/addons/exec.go:54]: done as suggested.

File: pkg/cmd/enable/addons/exec.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	names := strings.Split(o.names, ",")
+	for _, n := range names {
+		if _, ok := alreadyProvidedAddons[n]; !ok {
+			alreadyProvidedAddons[n] = true
+			addons = append(addons, strings.TrimSpace(n))
+		}
+	}
+	o.values.addons = addons
+
+	alreadyProvidedClusters := make(map[string]bool)
```
Comment (qiujian16) [inline pkg/cmd/enable/addons/exec.go:66]: this seems not used?

File: pkg/cmd/install/addons/cmd.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+			}
+			if err := o.run(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&o.names, "names", "", "Names of the built-in add-on to install (comma separated)")
```
Comment (qiujian16) [inline pkg/cmd/install/addons/cmd.go:48]: the builtin addons are: application-manager

**Review by qiujian16 — COMMENTED** [review]
nit, this looks good now. I think we need to indicate the valid built-in addon name in the help message. I think you would need to squash the commit.

/approve

File: pkg/cmd/enable/addons/exec.go (hunk 5 above)
Comment (mikeshng) [reply pkg/cmd/enable/addons/exec.go:66]: This is used to filter out duplicated cluster names. For example in the command below, `cluster1` is repeated 3 times but after filtering only 1 value is being used.
```
$ clusteradm enable addons --names application-manager --clusters cluster1,cluster1,cluster1 -v 3
I0720 11:29:14.378189   15685 exec.go:30] "addon options:" dry-run=false names="application-manager" clusters="cluster1,cluster1,cluster1" output-file=""
I0720 11:29:14.378239   15685 exec.go:77] "values:" addons=[application-manager] clusters=[cluster1]
```

Comment (qiujian16) [reply pkg/cmd/enable/addons/exec.go:66]: I see, I think you can use set.String.

**Pull request merged**